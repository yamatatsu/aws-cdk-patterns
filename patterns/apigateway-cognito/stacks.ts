import {
  App,
  Stack,
  StackProps,
  aws_lambda as lambda,
  aws_apigateway as apigateway,
  aws_cognito as cognito,
} from "aws-cdk-lib"

export class ApigatewayCognito extends Stack {
  constructor(parent: App, id: string, props?: StackProps) {
    super(parent, id, props)

    const userPool = new cognito.UserPool(this, "UserPool", {
      selfSignUpEnabled: true,
      /**
       * emailのみにすると SignUp API にて UsernameExistsException によるユーザー存在確認ができてしまう。
       * `username: true` を追加することで メール疎通以前においてメアド重複エラーが発生しなくなるため、ユーザー存在確認ができなくなる。
       * ただし username の入力が必須になる。変更はできないので変更機能の実装は不要。
       * 変更もしたい場合は preferred_username を使う。
       *
       * SignUp API の呼び出しサンプル
       *
       * curl https://cognito-idp.ap-northeast-1.amazonaws.com/ \
       *   -H 'Content-Type: application/x-amz-json-1.1' \
       *   -H 'x-amz-target: AWSCognitoIdentityProviderService.SignUp' \
       *   -d '{"ClientId":"xxxxxxxxxxxxxxxxx","Username":"hogehoge","UserAttributes": [{"Name":"email","Value":"xxxxxxxxxx@gmail.com"}],"Password": "xxxxxxxxxx","ValidationData":null}'
       *
       * curl https://cognito-idp.ap-northeast-1.amazonaws.com/ \
       *   -H 'Content-Type: application/x-amz-json-1.1' \
       *   -H 'x-amz-target: AWSCognitoIdentityProviderService.ConfirmSignUp' \
       *   -d '{"ClientId":"xxxxxxxxxxxxxxxxx","Username":"hogehoge","ConfirmationCode":"xxxxxx"}'
       */
      signInAliases: { username: true, email: true },
      userVerification: {
        emailStyle: cognito.VerificationEmailStyle.CODE,
      },
    })
    const userPoolClient = new cognito.UserPoolClient(this, "UserPoolClient", {
      userPool,
      authFlows: { userSrp: true },
    })

    new cognito.UserPoolDomain(this, "UserPoolDomain", {
      userPool,
      cognitoDomain: { domainPrefix: `yamatatsu-apigateway-example` },
    })

    const handler = new lambda.Function(this, "Lambda", {
      code: new lambda.AssetCode("./lambda"),
      handler: "index.handler",
      runtime: lambda.Runtime.NODEJS_14_X,
    })

    const restApi = new apigateway.RestApi(this, "RestApi", {
      restApiName: "ApigatewayCognito_RestApi",
      deploy: true,
    })

    const cognitoAuthrizer = new apigateway.CfnAuthorizer(
      this,
      "CfnAuthorizer",
      {
        name: "Cognito-Userpool",
        type: "COGNITO_USER_POOLS",
        restApiId: restApi.restApiId,
        identitySource: "method.request.header.x-ytoken",
        providerArns: [userPool.userPoolArn],
      },
    )

    restApi.root.addProxy({
      defaultIntegration: new apigateway.LambdaIntegration(handler),
      defaultMethodOptions: {
        authorizer: {
          authorizationType: apigateway.AuthorizationType.COGNITO,
          authorizerId: cognitoAuthrizer.ref,
        },
      },
    })
  }
}
