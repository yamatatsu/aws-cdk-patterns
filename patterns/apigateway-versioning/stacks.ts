import * as cdk from "@aws-cdk/core"
import * as lambda from "@aws-cdk/aws-lambda"
import * as apigateway from "@aws-cdk/aws-apigateway"

export class ApigatewayVersioning extends cdk.Stack {
  constructor(parent: cdk.App, id: string, props?: cdk.StackProps) {
    super(parent, id, props)

    const handler = new lambda.Function(this, "Lambda", {
      code: new lambda.AssetCode("./lambda"),
      handler: "index.handler",
      runtime: lambda.Runtime.NODEJS_12_X,
      tracing: lambda.Tracing.ACTIVE,
      currentVersionOptions: {
        removalPolicy: cdk.RemovalPolicy.RETAIN,
      },
    })

    const restApi = new apigateway.RestApi(this, "RestApi", {
      restApiName: "ApiGatewayVersioning",
      endpointTypes: [apigateway.EndpointType.REGIONAL],
      deployOptions: {
        loggingLevel: apigateway.MethodLoggingLevel.INFO,
      },
      defaultIntegration: new apigateway.LambdaIntegration(handler),
    })

    const fixVersion = genFixVersion(this, handler, restApi)
    fixVersion("v1.0.0", "6")
    fixVersion("v1.0.1", "7")
    fixVersion("v1.0.2", handler.currentVersion.version)
  }
}

const genFixVersion = (
  scope: cdk.Stack,
  handler: lambda.IFunction,
  restApi: apigateway.IRestApi,
) => (versionName: string, version: string) => {
  const _versionName = versionName.replace(/\./g, "_")

  const alias = lambda.Version.fromVersionAttributes(
    scope,
    `lambdaVersion_${_versionName}`,
    { lambda: handler, version },
  ).addAlias(_versionName)

  restApi.root.addResource(versionName).addProxy({
    defaultIntegration: new apigateway.LambdaIntegration(alias),
  })
}
