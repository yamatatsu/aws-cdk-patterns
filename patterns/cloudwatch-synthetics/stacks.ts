// FIXME: RCにはまだ synthetics が実装されていない。実装を待つ。
// import {
//   App,
//   Stack,
//   StackProps,
//   RemovalPolicy,
//   Duration,
//   aws_synthetics as synthetics,
//   aws_s3 as s3,
// } from "aws-cdk-lib"

// type Props = StackProps & {
//   code: synthetics.Code
// }

// export class CloudwatchSynthetics extends Stack {
//   constructor(parent: App, id: string, props: Props) {
//     super(parent, id, props)

//     const bucket = new s3.Bucket(this, "Bucket", {
//       removalPolicy: RemovalPolicy.DESTROY,
//     })

//     new synthetics.Canary(this, "Canary", {
//       artifactsBucketLocation: { bucket },
//       // role?: iam.IRole
//       timeToLive: Duration.minutes(5),
//       schedule: synthetics.Schedule.once(),
//       // startAfterCreation?: boolean
//       // successRetentionPeriod: Duration.days(7),
//       // failureRetentionPeriod: Duration.days(0),
//       // canaryName?: string
//       runtime: synthetics.Runtime.SYNTHETICS_NODEJS_PUPPETEER_3_0,
//       test: synthetics.Test.custom({
//         code: props.code,
//         handler: "index.handler",
//       }),
//     })
//   }
// }
