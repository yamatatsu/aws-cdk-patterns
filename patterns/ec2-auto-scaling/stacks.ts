import * as cdk from "@aws-cdk/core"
import * as iam from "@aws-cdk/aws-iam"
import * as ec2 from "@aws-cdk/aws-ec2"
import * as autoscaling from "@aws-cdk/aws-autoscaling"
import * as elb from "@aws-cdk/aws-elasticloadbalancingv2"

export class Ec2AutoScaling extends cdk.Stack {
  constructor(parent: cdk.App, id: string, props?: cdk.StackProps) {
    super(parent, id, props)

    const vpc = new ec2.Vpc(this, "Vpc", {
      maxAzs: 2,
      natGateways: 0,
    })

    // bastion
    // const bastion = new ec2.Instance(this, "Instance", {
    //   vpc,
    //   instanceType: ec2.InstanceType.of(
    //     ec2.InstanceClass.T2,
    //     ec2.InstanceSize.NANO,
    //   ),
    //   machineImage: ec2.MachineImage.latestAmazonLinux(),
    // })
    // // for ssm session manager
    // bastion.role.addManagedPolicy(
    //   iam.ManagedPolicy.fromManagedPolicyArn(
    //     this,
    //     "IAM-POLICY-AmazonSSMManagedInstanceCore",
    //     "arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore",
    //   ),
    // )

    const autoScalingGroup = new autoscaling.AutoScalingGroup(
      this,
      "AutoScalingGroup",
      {
        vpc,
        instanceType: ec2.InstanceType.of(
          ec2.InstanceClass.T2,
          ec2.InstanceSize.NANO,
        ),
        machineImage: ec2.MachineImage.latestAmazonLinux(),
      },
    )

    const alb = new elb.ApplicationLoadBalancer(this, "LoadBalancer", {
      vpc,
      internetFacing: true,
    })

    const listener = alb.addListener("Listener", {
      port: 80,
    })
    listener.addTargets("Target", {
      port: 80,
      targets: [autoScalingGroup],
    })

    autoScalingGroup.scaleOnRequestCount("AModestLoad", {
      targetRequestsPerSecond: 1,
    })
    listener.connections.allowDefaultPortFromAnyIpv4("Open to the world")
  }
}
