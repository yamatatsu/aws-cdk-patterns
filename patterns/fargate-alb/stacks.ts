import path from "path"
import {
  App,
  Stack,
  StackProps,
  aws_ec2 as ec2,
  aws_elasticloadbalancingv2 as elb,
  aws_ecs as ecs,
  aws_iam as iam,
} from "aws-cdk-lib"

export class FargateAlb extends Stack {
  constructor(parent: App, id: string, props?: StackProps) {
    super(parent, id, props)

    const vpc = new ec2.Vpc(this, "Vpc", {
      maxAzs: 2,
      natGatewayProvider: ec2.NatInstanceProvider.instance({
        instanceType: new ec2.InstanceType("t2.nano"),
      }),
    })

    const cluster = new ecs.Cluster(this, "Cluster", {
      vpc,
    })

    // FargateTaskDefinition のデフォルトのRoleではlogsの権限が足りなくてログが吐けないため、作成する
    const executionRole = new iam.Role(this, "EcsTaskExecutionRole", {
      roleName: "ecs-task-execution-role",
      assumedBy: new iam.ServicePrincipal("ecs-tasks.amazonaws.com"),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName(
          "service-role/AmazonECSTaskExecutionRolePolicy",
        ),
      ],
    })

    const taskDefinition = new ecs.FargateTaskDefinition(
      this,
      "FargateTaskDefinition",
      { executionRole },
    )

    const container = taskDefinition.addContainer("webContainer", {
      image: ecs.ContainerImage.fromAsset(path.resolve(__dirname, "server")),
      // コンテナのログがCloudWatchに吐かれるようになる
      logging: new ecs.AwsLogDriver({
        streamPrefix: "fargate-task",
      }),
      portMappings: [{ containerPort: 8000 }],
    })

    const service = new ecs.FargateService(this, "FargateService", {
      cluster,
      taskDefinition,
    })

    const loadBalancer = new elb.ApplicationLoadBalancer(this, "LB", {
      vpc,
      internetFacing: true,
    })
    const listener = loadBalancer.addListener("PublicListener", {
      protocol: elb.ApplicationProtocol.HTTP,
    })
    listener.addTargets("ECS", {
      protocol: elb.ApplicationProtocol.HTTP,
      // TODO: serviceに複数のweb serverがあるときにどんな挙動をするのか？
      targets: [service],
    })
  }
}
