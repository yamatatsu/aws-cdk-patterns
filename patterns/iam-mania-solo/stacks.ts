import { App, aws_iam, Stack, StackProps } from "aws-cdk-lib"

export class IamManiaSolo extends Stack {
  constructor(parent: App, id: string, props?: StackProps) {
    super(parent, id, props)

    const mfaForcePolicy = new aws_iam.ManagedPolicy(this, "MFAForcePolicy", {
      managedPolicyName: "MFAForcePolicy",
      statements: [
        // [DENY]MFA必須化
        new aws_iam.PolicyStatement({
          effect: aws_iam.Effect.DENY,
          notActions: ["iam:*"],
          resources: ["*"],
          conditions: {
            BoolIfExists: { "aws:MultiFactorAuthPresent": "false" },
          },
        }),
        // [ALLOW]全ユーザーのリスト表示
        new aws_iam.PolicyStatement({
          effect: aws_iam.Effect.ALLOW,
          actions: ["iam:ListUsers", "iam:ListVirtualMFADevices"],
          resources: ["*"],
        }),
      ],
    })

    const adminGroup = new aws_iam.Group(this, "AdminGroup", {
      groupName: "AdminGroup",
      managedPolicies: [
        aws_iam.ManagedPolicy.fromAwsManagedPolicyName("ReadOnlyAccess"),
        mfaForcePolicy,
      ],
    })

    const soloUserName = this.node.tryGetContext("soloUserName")
    const user = aws_iam.User.fromUserName(this, "SoloUser", soloUserName)

    user.addToGroup(adminGroup)

    new aws_iam.Role(this, "AdminRole", {
      roleName: "AdminRole",
      assumedBy: new aws_iam.ArnPrincipal(user.userArn),
      managedPolicies: [
        aws_iam.ManagedPolicy.fromAwsManagedPolicyName("AdministratorAccess"),
      ],
    })
  }
}
