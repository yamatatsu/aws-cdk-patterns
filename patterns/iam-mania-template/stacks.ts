import { App, Arn, aws_iam, Stack, StackProps } from "aws-cdk-lib"

export class IamManiaAdmin extends Stack {
  constructor(parent: App, id: string, props?: StackProps) {
    super(parent, id, props)

    const ipRestrictionAndMFAForcePolicy = new aws_iam.ManagedPolicy(
      this,
      "IpRestrictionAndMFAForcePolicy",
      {
        managedPolicyName: "IpRestrictionAndMFAForcePolicy",
        statements: [
          // [DENY]IP制限
          new aws_iam.PolicyStatement({
            effect: aws_iam.Effect.DENY,
            actions: ["*"],
            resources: ["*"],
            // 任意のIPに変更すること
            conditions: { NotIpAddress: { "aws:SourceIp": ["8.8.8.8/32"] } },
          }),
          // [DENY]MFA必須化
          new aws_iam.PolicyStatement({
            effect: aws_iam.Effect.DENY,
            notActions: ["iam:*"],
            resources: ["*"],
            conditions: {
              BoolIfExists: { "aws:MultiFactorAuthPresent": "false" },
            },
          }),
          // [ALLOW]全ユーザーのリスト表示
          new aws_iam.PolicyStatement({
            effect: aws_iam.Effect.ALLOW,
            actions: ["iam:ListUsers", "iam:ListVirtualMFADevices"],
            resources: ["*"],
          }),
        ],
      },
    )
    const changeOwnPasswordPolicy = new aws_iam.ManagedPolicy(
      this,
      "ChangeOwnPasswordPolicy",
      {
        managedPolicyName: "ChangeOwnPasswordPolicy",
        statements: [
          // [ALLOW]自身のパスワード変更＆MFA有効化ポリシー
          new aws_iam.PolicyStatement({
            effect: aws_iam.Effect.ALLOW,
            actions: [
              "iam:ChangePassword",
              "iam:CreateAccessKey",
              "iam:CreateVirtualMFADevice",
              "iam:DeactivateMFADevice",
              "iam:DeleteAccessKey",
              "iam:DeleteVirtualMFADevice",
              "iam:EnableMFADevice",
              "iam:GetAccountPasswordPolicy",
              "iam:UpdateAccessKey",
              "iam:UpdateSigningCertificate",
              "iam:UploadSigningCertificate",
              "iam:UpdateLoginProfile",
              "iam:ResyncMFADevice",
            ],
            resources: [
              Arn.format(
                {
                  service: "iam",
                  region: "",
                  resource: "user",
                  resourceName: "${aws:username}",
                },
                this,
              ),
              Arn.format(
                {
                  service: "iam",
                  region: "",
                  resource: "mfa",
                  resourceName: "${aws:username}",
                },
                this,
              ),
            ],
          }),
          // [ALLOW]全ユーザーのリスト表示
          new aws_iam.PolicyStatement({
            effect: aws_iam.Effect.ALLOW,
            actions: ["iam:ListUsers", "iam:ListVirtualMFADevices"],
            resources: ["*"],
          }),
        ],
      },
    )
    const denyNetworkAccessPolicy = new aws_iam.ManagedPolicy(
      this,
      "DenyNetworkAccessPolicy",
      {
        managedPolicyName: "DenyNetworkAccessPolicy",
        statements: [
          new aws_iam.PolicyStatement({
            effect: aws_iam.Effect.DENY,
            actions: [
              "ec2:AcceptVpcEndpointConnections",
              "ec2:AssociateDhcpOptions",
              "ec2:AssociateRouteTable",
              "ec2:AssociateSubnetCidrBlock",
              "ec2:AssociateVpcCidrBlock",
              "ec2:AttachInternetGateway",
              "ec2:AttachNetworkInterface",
              "ec2:AttachVpnGateway",
              "ec2:CreateCustomerGateway",
              "ec2:CreateDefaultSubnet",
              "ec2:CreateDefaultVpc",
              "ec2:CreateDhcpOptions",
              "ec2:CreateEgressOnlyInternetGateway",
              "ec2:CreateInternetGateway",
              "ec2:CreateNetworkAcl",
              "ec2:CreateNetworkAclEntry",
              "ec2:CreateRoute",
              "ec2:CreateRouteTable",
              "ec2:CreateSecurityGroup",
              "ec2:CreateSubnet",
              "ec2:CreateVpc",
              "ec2:CreateVpcEndpoint",
              "ec2:CreateVpcEndpointConnectionNotification",
              "ec2:CreateVpcEndpointServiceConfiguration",
              "ec2:CreateVpnConnection",
              "ec2:CreateVpnConnectionRoute",
              "ec2:CreateVpnGateway",
              "ec2:DeleteEgressOnlyInternetGateway",
              "ec2:DeleteNatGateway",
              "ec2:DeleteNetworkInterface",
              "ec2:DeleteNetworkInterfacePermission",
              "ec2:DeletePlacementGroup",
              "ec2:DeleteSubnet",
              "ec2:DeleteVpc",
              "ec2:DeleteVpcEndpointConnectionNotifications",
              "ec2:DeleteVpcEndpoints",
              "ec2:DeleteVpcEndpointServiceConfigurations",
              "ec2:DeleteVpnConnection",
              "ec2:DeleteVpnConnectionRoute",
              "ec2:DeleteVpnGateway",
              "ec2:DetachInternetGateway",
              "ec2:DetachNetworkInterface",
              "ec2:DetachVpnGateway",
              "ec2:DisableVgwRoutePropagation",
              "ec2:DisableVpcClassicLinkDnsSupport",
              "ec2:DisassociateRouteTable",
              "ec2:DisassociateSubnetCidrBlock",
              "ec2:DisassociateVpcCidrBlock",
              "ec2:EnableVgwRoutePropagation",
              "ec2:EnableVpcClassicLinkDnsSupport",
              "ec2:ModifyNetworkInterfaceAttribute",
              "ec2:ModifySubnetAttribute",
              "ec2:ModifyVpcAttribute",
              "ec2:ModifyVpcEndpoint",
              "ec2:ModifyVpcEndpointConnectionNotification",
              "ec2:ModifyVpcEndpointServiceConfiguration",
              "ec2:ModifyVpcEndpointServicePermissions",
              "ec2:ModifyVpcPeeringConnectionOptions",
              "ec2:ModifyVpcTenancy",
              "ec2:MoveAddressToVpc",
              "ec2:RejectVpcEndpointConnections",
              "ec2:ReplaceNetworkAclAssociation",
              "ec2:ReplaceNetworkAclEntry",
              "ec2:ReplaceRoute",
              "ec2:ReplaceRouteTableAssociation",
              "ec2:ResetNetworkInterfaceAttribute",
              "ec2:RestoreAddressToClassic",
              "ec2:UpdateSecurityGroupRuleDescriptionsEgress",
              "ec2:UpdateSecurityGroupRuleDescriptionsIngress",
              "directconnect:*",
            ],
            resources: ["*"],
          }),
          new aws_iam.PolicyStatement({
            effect: aws_iam.Effect.DENY,
            actions: [
              "ec2:AcceptVpcPeeringConnection",
              "ec2:AttachClassicLinkVpc",
              "ec2:AuthorizeSecurityGroupEgress",
              "ec2:AuthorizeSecurityGroupIngress",
              "ec2:CreateVpcPeeringConnection",
              "ec2:DeleteCustomerGateway",
              "ec2:DeleteDhcpOptions",
              "ec2:DeleteInternetGateway",
              "ec2:DeleteNetworkAcl",
              "ec2:DeleteNetworkAclEntry",
              "ec2:DeleteRoute",
              "ec2:DeleteRouteTable",
              "ec2:DeleteSecurityGroup",
              "ec2:DeleteVolume",
              "ec2:DeleteVpcPeeringConnection",
              "ec2:DetachClassicLinkVpc",
              "ec2:DisableVpcClassicLink",
              "ec2:EnableVpcClassicLink",
              "ec2:GetConsoleScreenshot",
              "ec2:RejectVpcPeeringConnection",
              "ec2:RevokeSecurityGroupEgress",
              "ec2:RevokeSecurityGroupIngress",
              "ec2:AcceptTransitGatewayVpcAttachment",
              "ec2:AssociateTransitGatewayRouteTable",
              "ec2:CreateTransitGateway",
              "ec2:CreateTransitGatewayRoute",
              "ec2:CreateTransitGatewayRouteTable",
              "ec2:CreateTransitGatewayVpcAttachment",
              "ec2:DeleteTransitGateway",
              "ec2:DeleteTransitGatewayRoute",
              "ec2:DeleteTransitGatewayRouteTable",
              "ec2:DeleteTransitGatewayVpcAttachment",
              "ec2:DescribeTransitGatewayAttachments",
              "ec2:DescribeTransitGatewayRouteTables",
              "ec2:DescribeTransitGatewayVpcAttachments",
              "ec2:DescribeTransitGateways",
              "ec2:DisableTransitGatewayRouteTablePropagation",
              "ec2:DisassociateTransitGatewayRouteTable",
              "ec2:EnableTransitGatewayRouteTablePropagation",
              "ec2:ExportTransitGatewayRoutes",
              "ec2:GetTransitGatewayAttachmentPropagations",
              "ec2:GetTransitGatewayRouteTableAssociations",
              "ec2:GetTransitGatewayRouteTablePropagations",
              "ec2:ModifyTransitGatewayVpcAttachment",
              "ec2:RejectTransitGatewayVpcAttachment",
              "ec2:ReplaceTransitGatewayRoute",
              "ec2:SearchTransitGatewayRoutes",
            ],
            resources: ["*"],
          }),
        ],
      },
    )

    const adminGroup = new aws_iam.Group(this, "AdminGroup", {
      groupName: "Admins",
      managedPolicies: [
        aws_iam.ManagedPolicy.fromAwsManagedPolicyName("AdministratorAccess"),
        ipRestrictionAndMFAForcePolicy,
      ],
    })
    const networkAdminGroup = new aws_iam.Group(this, "NetworkAdminGroup", {
      groupName: "NetworkAdmins",
      managedPolicies: [
        aws_iam.ManagedPolicy.fromAwsManagedPolicyName(
          "job-function/NetworkAdministrator",
        ),
        aws_iam.ManagedPolicy.fromAwsManagedPolicyName("ReadOnlyAccess"),
        ipRestrictionAndMFAForcePolicy,
        changeOwnPasswordPolicy,
      ],
    })
    const developerGroup = new aws_iam.Group(this, "DeveloperGroup", {
      groupName: "Developers",
      managedPolicies: [
        aws_iam.ManagedPolicy.fromAwsManagedPolicyName("PowerUserAccess"),
        ipRestrictionAndMFAForcePolicy,
        changeOwnPasswordPolicy,
        denyNetworkAccessPolicy,
      ],
    })
    const operatorGroup = new aws_iam.Group(this, "OperatorGroup", {
      groupName: "Operators",
      managedPolicies: [
        aws_iam.ManagedPolicy.fromAwsManagedPolicyName("AWSSupportAccess"),
        aws_iam.ManagedPolicy.fromAwsManagedPolicyName("ReadOnlyAccess"),
        ipRestrictionAndMFAForcePolicy,
        changeOwnPasswordPolicy,
      ],
    })
    const billingUserGroup = new aws_iam.Group(this, "BillingUserGroup", {
      groupName: "BillingUsers",
      managedPolicies: [
        aws_iam.ManagedPolicy.fromAwsManagedPolicyName("job-function/Billing"),
        aws_iam.ManagedPolicy.fromAwsManagedPolicyName("ReadOnlyAccess"),
        ipRestrictionAndMFAForcePolicy,
        changeOwnPasswordPolicy,
      ],
    })

    const user = new aws_iam.User(this, "user", {
      userName: "hoge",
      groups: [adminGroup],
    })

    new aws_iam.Role(this, "AdminRole", {
      roleName: "AdminRole",
      assumedBy: new aws_iam.ArnPrincipal(user.userArn),
      managedPolicies: [
        aws_iam.ManagedPolicy.fromAwsManagedPolicyName("AdministratorAccess"),
      ],
    })
  }
}
