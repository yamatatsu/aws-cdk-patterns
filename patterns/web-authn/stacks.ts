import * as cdk from "@aws-cdk/core"
import * as lambda from "@aws-cdk/aws-lambda"
import * as apigateway from "@aws-cdk/aws-apigatewayv2"
import * as s3 from "@aws-cdk/aws-s3"
import * as s3Deployment from "@aws-cdk/aws-s3-deployment"
import * as cloudfront from "@aws-cdk/aws-cloudfront"
import { LambdaProxyIntegration } from "@aws-cdk/aws-apigatewayv2"

export class WebAuthnDemo extends cdk.Stack {
  constructor(parent: cdk.App, id: string, props?: cdk.StackProps) {
    super(parent, id, props)

    const code = lambda.Code.fromAsset("./dist")
    const getChallengeHandler = new lambda.Function(
      this,
      "getChallengeHandler",
      {
        functionName: "WebAuthnDemo_getChallengeHandler",
        runtime: lambda.Runtime.NODEJS_12_X,
        code,
        handler: "index.getChallenge",
        timeout: cdk.Duration.seconds(3),
        memorySize: 128,
        tracing: lambda.Tracing.ACTIVE,
      },
    )

    const api = new apigateway.HttpApi(this, "HttpApi", {
      apiName: "WebAuthnDemoApi",
      corsPreflight: {
        // allowOrigins?: string[];
      },
    })

    api.addRoutes({
      path: "/auth/challenge",
      methods: [apigateway.HttpMethod.GET],
      integration: new LambdaProxyIntegration({
        handler: getChallengeHandler,
      }),
    })
  }
}
