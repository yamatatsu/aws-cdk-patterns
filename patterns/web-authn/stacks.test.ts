import { SynthUtils } from "@aws-cdk/assert"
import * as cdk from "@aws-cdk/core"
import { WebAuthnDemo } from "./stacks"

test("snapshot test", () => {
  const app = new cdk.App()

  const target = new WebAuthnDemo(app, "Target")

  expect(SynthUtils.toCloudFormation(target)).toMatchSnapshot()
})
