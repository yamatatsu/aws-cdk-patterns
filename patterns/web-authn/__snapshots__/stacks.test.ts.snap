// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters5b385fa3c561a46cab5a75fb5516f08be77c83c92604538d09437216e61c7899ArtifactHash3A4B6B2F": Object {
      "Description": "Artifact hash for asset \\"5b385fa3c561a46cab5a75fb5516f08be77c83c92604538d09437216e61c7899\\"",
      "Type": "String",
    },
    "AssetParameters5b385fa3c561a46cab5a75fb5516f08be77c83c92604538d09437216e61c7899S3Bucket19DD2848": Object {
      "Description": "S3 bucket for asset \\"5b385fa3c561a46cab5a75fb5516f08be77c83c92604538d09437216e61c7899\\"",
      "Type": "String",
    },
    "AssetParameters5b385fa3c561a46cab5a75fb5516f08be77c83c92604538d09437216e61c7899S3VersionKey83E6A27D": Object {
      "Description": "S3 key for asset version \\"5b385fa3c561a46cab5a75fb5516f08be77c83c92604538d09437216e61c7899\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "HttpApiDefaultStage3EEB07D6": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "HttpApiF5A9A8A7",
        },
        "AutoDeploy": true,
        "StageName": "$default",
      },
      "Type": "AWS::ApiGatewayV2::Stage",
    },
    "HttpApiF5A9A8A7": Object {
      "Properties": Object {
        "CorsConfiguration": Object {},
        "Name": "WebAuthnDemoApi",
        "ProtocolType": "HTTP",
      },
      "Type": "AWS::ApiGatewayV2::Api",
    },
    "HttpApiGETauthchallenge573B056E": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "HttpApiF5A9A8A7",
        },
        "RouteKey": "GET /auth/challenge",
        "Target": Object {
          "Fn::Join": Array [
            "",
            Array [
              "integrations/",
              Object {
                "Ref": "HttpApiGETauthchallengeGETauthchallengeIntegrationBDA6710B",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "HttpApiGETauthchallengeGETauthchallengeIntegrationBDA6710B": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "HttpApiF5A9A8A7",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": Object {
          "Fn::GetAtt": Array [
            "getChallengeHandler8D5C0935",
            "Arn",
          ],
        },
        "PayloadFormatVersion": "2.0",
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "HttpApiGETauthchallengeTargetHttpApiGETauthchallenge682056F5PermissionA2BFF7DC": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "getChallengeHandler8D5C0935",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "HttpApiF5A9A8A7",
              },
              "/*/*/auth/challenge",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "getChallengeHandler8D5C0935": Object {
      "DependsOn": Array [
        "getChallengeHandlerServiceRoleDefaultPolicy0BAD431D",
        "getChallengeHandlerServiceRole0F75F0BF",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters5b385fa3c561a46cab5a75fb5516f08be77c83c92604538d09437216e61c7899S3Bucket19DD2848",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters5b385fa3c561a46cab5a75fb5516f08be77c83c92604538d09437216e61c7899S3VersionKey83E6A27D",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters5b385fa3c561a46cab5a75fb5516f08be77c83c92604538d09437216e61c7899S3VersionKey83E6A27D",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "FunctionName": "WebAuthnDemo_getChallengeHandler",
        "Handler": "index.getChallenge",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "getChallengeHandlerServiceRole0F75F0BF",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 3,
        "TracingConfig": Object {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "getChallengeHandlerServiceRole0F75F0BF": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "getChallengeHandlerServiceRoleDefaultPolicy0BAD431D": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "getChallengeHandlerServiceRoleDefaultPolicy0BAD431D",
        "Roles": Array [
          Object {
            "Ref": "getChallengeHandlerServiceRole0F75F0BF",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
